### PR TITLE
Correct parent chain in v2 push when v1Compatibility files on the disk are inconsistent

### PR DESCRIPTION
**Pull request by aaronlehmann**
This fixes an issue where two images with the same filesystem contents
and configuration but different remote IDs could share a v1Compatibility
file, resulting in corrupted manifests.

It would be great to get this into 1.9.1 if possible.

@tonistiigi @dmcgowan @tiborvass

### CODE REVIEW

**Comment by tonistiigi** [comment]
LGTM

File: graph/push_v2.go [hunk 1]
```diff
@@ -226,6 +233,74 @@ func (p *v2Pusher) pushV2Tag(tag string) error {
...
+func fixHistory(m *schema1.Manifest) error {
+	var lastID string
+
+	for i := len(m.History) - 1; i >= 0; i-- {
+		var historyEntry map[string]*json.RawMessage
+		if err := json.Unmarshal([]byte(m.History[i].V1Compatibility), &historyEntry); err != nil {
+			return err
+		}
+
+		idJSON, present := historyEntry["id"]
```
Comment (LK4D4) [inline graph/push_v2.go]: Now I think maybe it's better to use temp type for unmarshaling, map looks weird.

File: graph/push_v2.go [hunk 2]
```diff
@@ -226,6 +233,74 @@ func (p *v2Pusher) pushV2Tag(tag string) error {
...
+				if parent != "" {
+					logrus.Debugf("parent id mismatch detected; fixing. parent reference: %s", parent)
+					delete(historyEntry, "parent")
+					fixedHistory, err := json.Marshal(historyEntry)
+					if err != nil {
+						return err
+					}
+					m.History[i].V1Compatibility = string(fixedHistory)
+				}
+			}
```
Comment (LK4D4) [inline graph/push_v2.go]: ah, I see, it's not possible

File: graph/push_v2.go (hunk 1 above)
Comment (crosbymichael) [inline graph/push_v2.go]: ya, why unmarshal twice when you can just use a temp type of something?

Comment (LK4D4) [inline graph/push_v2.go]: Maybe I'm missing something, but this looks just like 
```
type X struct {
    ID string
    Parent string
}
```

Comment (tonistiigi) [inline graph/push_v2.go]: The `historyEntry` is not only parsed but needs to be Marshalled back later with the changes.

Comment (LK4D4) [inline graph/push_v2.go]: I don't see contradiction with typed struct.

Comment (aaronlehmann) [inline graph/push_v2.go]: We need to be able to handle configurations from future versions that may include fields we don't know about.

Comment (LK4D4) [inline graph/push_v2.go]: Hmm, don't see contradiction either :)

Comment (LK4D4) [inline graph/push_v2.go]: Oooh, I got it, you need to change only specified field and write the rest.

File: graph/push_v2.go [hunk 3]
```diff
@@ -226,6 +233,74 @@ func (p *v2Pusher) pushV2Tag(tag string) error {
 	return manSvc.Put(signed)
 }
 
+func fixHistory(m *schema1.Manifest) error {
```
Comment (LK4D4) [inline graph/push_v2.go]: Could you pls write here comment why we need this.

**Comment by aaronlehmann** [comment]
Updated with more comments.

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**